nomos-services/data-availability/sampling/src/lib.rs

```diff
@@ -287,61 +299,155 @@ where
+                Self::handle_error(error, sampler, network_adapter, verifier).await
             }
             SamplingEvent::SamplingRequest {
                 blob_id,
                 share_idx,
                 response_sender,
             } => {
-                info_with_id!(blob_id, "SamplingRequest");
-                let maybe_share = storage_adapter
-                    .get_light_share(blob_id, share_idx.to_le_bytes())
-                    .await
-                    .map_err(|error| {
-                        error!("Failed to get share from storage adapter: {error}");
-                        None::<SamplingBackend::Share>
-                    })
-                    .ok()
-                    .flatten();
-
-                if response_sender.send(maybe_share).await.is_err() {
-                    error!("Error sending sampling response");
-                }
+                Self::handle_request(storage_adapter, blob_id, share_idx, response_sender).await;
+                None
             }
         }
     }
 
+    async fn handle_success(
+        sampler: &mut SamplingBackend,
+        verifier: &ShareVerifier,
+        blob_id: BlobId,
+        light_share: Box<DaLightShare>,
+    ) {
+        info_with_id!(blob_id, "SamplingSuccess");
+
+        let Some(commitments) = sampler.get_commitments(&blob_id) else {
+            error_with_id!(blob_id, "Error getting commitments for blob");
+            sampler.handle_sampling_error(blob_id).await;
+            return;
+        };
+
+        if verifier.verify(&commitments, &light_share).is_err() {
+            error_with_id!(blob_id, "SamplingError");
+            sampler.handle_sampling_error(blob_id).await;
+            return;
+        }
+
+        sampler
+            .handle_sampling_success(blob_id, light_share.share_idx)
+            .await;
+    }
+
+    async fn handle_error(
+        error: SamplingError,
+        sampler: &mut SamplingBackend,
+        network_adapter: &SamplingNetwork,
+        verifier: &ShareVerifier,
+    ) -> Option<(
+        BoxFuture<'static, ()>,
+        BoxFuture<'static, HistoricFallbackResult>,
+    )> {
+        let Some(blob_id) = error.blob_id() else {
+            error!("Error while sampling: {error}");
+            return None;
+        };
+
+        error_with_id!(blob_id, "SamplingError");
+
+        match error {
+            SamplingError::BlobNotFound { .. } => {
+                sampler.handle_sampling_error(*blob_id).await;
+            }
+            SamplingError::MismatchSession { blob_id, session } => {
+                return Self::handle_session_mismatch(
+                    blob_id,
+                    session,
+                    sampler,
+                    network_adapter,
+                    verifier,
+                )
+                .await;
+            }
+            _ => {
+                error!("Error while sampling: {error}");
+            }
+        }
```

Comment:
We should log the session mismatch and the retry here with the blob id and the session itself. For sanity on debugging.